include/secp256k1_musig.h

```diff
@@ -0,0 +1,428 @@
+ * session ID to ensure uniqueness.
+ *
+ * Fields:
+ *      combined_pk: MuSig-computed combined public key
+ *        n_signers: Number of signers
+ *          pk_hash: The 32-byte hash of the original public keys
+ *   combined_nonce: Summed combined public nonce (undefined if `nonce_is_set` is false)
+ *     nonce_is_set: Whether the above nonce has been set
+ * nonce_is_negated: If `nonce_is_set`, whether the above nonce was negated after
+ *                   summing the participants' nonces. Needed to ensure the nonce's y
+ *                   coordinate has a quadratic-residue y coordinate
+ *              msg: The 32-byte message (hash) to be signed
+ *       msg_is_set: Whether the above message has been set
+ *  has_secret_data: Whether this session object has a signers' secret data; if this
+ *                   is `false`, it may still be used for verification purposes.
+ *           seckey: If `has_secret_data`, the signer's secret key
+ *         secnonce: If `has_secret_data`, the signer's secret nonce
+ *            nonce: If `has_secret_data`, the signer's public nonce
+ *        nonce_commitments_hash: If `has_secret_data` and
+ *                   `nonce_commitments_hash_is_set`, the hash of all signers'
+ *                   commitments
+ * nonce_commitments_hash_is_set: If `has_secret_data`, whether the
+ *                   nonce_commitments_hash has been set
+ */
+typedef struct {
+    secp256k1_pubkey combined_pk;
+    size_t n_signers;
+    unsigned char pk_hash[32];
+    secp256k1_pubkey combined_nonce;
+    int nonce_is_set;
+    int nonce_is_negated;
+    unsigned char msg[32];
+    int msg_is_set;
+    int has_secret_data;
+    unsigned char seckey[32];
+    unsigned char secnonce[32];
+    secp256k1_pubkey nonce;
+    unsigned char nonce_commitments_hash[32];
+    int nonce_commitments_hash_is_set;
+} secp256k1_musig_session;
+
+/** Data structure containing data on all signers in a single session.
+ *
+ * The workflow for this structure is as follows:
+ *
+ * 1. This structure is initialized with `musig_session_initialize` or
+ *    `musig_session_initialize_verifier`, which set the `pubkey` and `index` fields,
+ *    and zero out all other fields. The public session is initialized with the
+ *    signers' nonce_commitments.
+ *
+ * 2. In a non-public session the nonce_commitments are set with the function
+ *    `musig_get_public_nonce`, which as a also returns the signer's public nonce.
+ *    This ensures that the public nonce is not exposed until all commitments have
+ *    been received.
+ *
+ * 3. Each individual data struct should be updated with `musig_set_nonce` once a
+ *    nonce is available. This function takes a single signer data struct rather than
+ *    an array because it may fail in the case that the provided nonce does not match
+ *    the commitment. In this case, it is desirable to identify the exact party whose
+ *    nonce was inconsistent.
+ *
```

Comment:
Maybe add that a session struct can be reused for a second session if it's initilaized again.

Reply:
yes if we memset it to 0